internal/broker/bind.go

```diff
@@ -3,36 +3,329 @@ package broker
+	instanceGetter          instanceGetter
+	bindTemplateRenderer    bindTemplateRenderer
+	bindTemplateResolver    bindTemplateResolver
+	instanceBindDataStorage instanceBindDataStorage
+	bindStateGetter         bindStateBindingGetter
+	bindOperationStorage    bindOperationStorage
+	operationIDProvider     func() (internal.OperationID, error)
+	mu                      sync.Mutex
+
+	log *logrus.Entry
+
+	testHookAsyncCalled func(internal.OperationID)
 }
 
-func (svc *bindService) Bind(ctx context.Context, osbCtx OsbContext, req *osb.BindRequest) (*osb.BindResponse, error) {
+func (svc *bindService) Bind(ctx context.Context, osbCtx OsbContext, req *osb.BindRequest) (*osb.BindResponse, *osb.HTTPStatusCodeError) {
 	if len(req.Parameters) > 0 {
-		return nil, fmt.Errorf("helm-broker does not support configuration options for the service binding")
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr("helm-broker does not support configuration options for the service binding")}
+	}
+
+	if !req.AcceptsIncomplete {
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr("asynchronous operation mode required")}
+	}
+
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	iID := internal.InstanceID(req.InstanceID)
+	bID := internal.BindingID(req.BindingID)
+	svcID := internal.ServiceID(req.ServiceID)
+	svcPlanID := internal.ServicePlanID(req.PlanID)
+	paramHash := jsonhash.HashS(req.Parameters)
+
+	switch opIDInProgress, inProgress, err := svc.bindStateGetter.IsBindingInProgress(iID, bID); true {
+	case err != nil:
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusInternalServerError, ErrorMessage: strPtr(fmt.Sprintf("while checking if service binding is being created: %v", err))}
+	case inProgress:
+		opKeyInProgress := osb.OperationKey(opIDInProgress)
+		return &osb.BindResponse{Async: true, OperationKey: &opKeyInProgress}, nil
+	}
+
+	switch bindOp, state, err := svc.bindStateGetter.IsBound(iID, bID); true {
+	case err != nil:
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusInternalServerError, ErrorMessage: strPtr(fmt.Sprintf("while checking if service binding for service instance already exists: %v", err))}
+	case state:
+		opID := bindOp.OperationID
+		bindInput, err := svc.prepareBindInput(osbCtx, iID, bID, svcID, svcPlanID, opID, paramHash)
+		if err != nil {
+			return nil, err
+		}
+
+		svc.doAsync(ctx, bindInput)
+		out, getIbdErr := svc.getInstanceBindData(iID, bID)
+		if getIbdErr != nil {
+			return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr(fmt.Sprintf("while getting bind data from memory for instance id: %q and service binding id: %q with error: %v", iID, bID, err))}
+		}
+
+		credsOut := svc.dtoFromModel(out.Credentials)
+
+		if err := svc.instanceBindDataStorage.Remove(iID); err != nil {
+			svc.log.Errorf("while removing instance bind data after getting it from memory on bind, got error: %v", err)
```

Comment:
why do you assume thedata is stored in memory? Please replace all messages with "from memory" with proper message.